ethers-providers/src/transports/quorum.rs

```diff
@@ -0,0 +1,640 @@
+    All,
+    /// The quorum is reached when the majority of the providers have returned a
+    /// matching value, taking into account their weight.
+    Majority,
+    /// The quorum is reached when the cumulative weight of a matching return
+    /// exceeds the given percentage of the total weight.
+    ///
+    /// NOTE: this must be less than `100u8`
+    Percentage(u8),
+    /// The quorum is reached when the given number of provider agree
+    /// The configured weight is ignored in this case.
+    ProviderCount(usize),
+    /// The quorum is reached once the accumulated weight of the matching return
+    /// exceeds this weight.
+    Weight(u64),
+}
+
+impl Quorum {
+    fn weight<T>(self, providers: &[WeightedProvider<T>]) -> u64 {
+        match self {
+            Quorum::All => providers.iter().map(|p| p.weight).sum::<u64>(),
+            Quorum::Majority => {
+                let total = providers.iter().map(|p| p.weight).sum::<u64>();
+                let rem = total % 2;
+                total / 2 + rem
+            }
+            Quorum::Percentage(p) => {
+                providers.iter().map(|p| p.weight).sum::<u64>() * (p as u64) / 100
+            }
+            Quorum::ProviderCount(num) => {
+                // take the lowest `num` weights
+                let mut weights = providers.iter().map(|p| p.weight).collect::<Vec<_>>();
+                weights.sort_unstable();
+                weights.into_iter().take(num).sum()
+            }
+            Quorum::Weight(w) => w,
+        }
+    }
+}
+
+impl Default for Quorum {
+    fn default() -> Self {
+        Quorum::Majority
+    }
+}
+
+// A future that returns the provider's response and it's index within the
+// `QuorumProvider` provider set
+#[cfg(target_arch = "wasm32")]
+type PendingRequest<'a> = Pin<Box<dyn Future<Output = (Result<Value, ProviderError>, usize)> + 'a>>;
+#[cfg(not(target_arch = "wasm32"))]
+type PendingRequest<'a> =
+    Pin<Box<dyn Future<Output = (Result<Value, ProviderError>, usize)> + 'a + Send>>;
+
+/// A future that only returns a value of the `QuorumProvider`'s provider
+/// reached a quorum.
+struct QuorumRequest<'a, T> {
+    inner: &'a QuorumProvider<T>,
+    /// The different answers with their cumulative weight
+    responses: Vec<(Value, u64)>,
+    /// All the errors the provider yielded
+    errors: Vec<ProviderError>,
+    // Requests currently pending
+    requests: Vec<PendingRequest<'a>>,
+}
+
+impl<'a, T> QuorumRequest<'a, T> {
+    fn new(inner: &'a QuorumProvider<T>, requests: Vec<PendingRequest<'a>>) -> Self {
+        Self {
+            responses: Vec::new(),
+            errors: Vec::new(),
+            inner,
+            requests,
+        }
+    }
+}
+
+impl<'a, T> Future for QuorumRequest<'a, T> {
+    type Output = Result<Value, QuorumError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        for n in (0..this.requests.len()).rev() {
```

Comment:
Why are we iterating in reverse here?

Reply:
because how swap_remove works, which replaces the removed element with the last element. but this also means that we start at `0` we would miss the value the last index, therefor we need to start at the last index.